### PR TITLE
 implement group by related object field support and introduce new tests

### LINKED ISSUE #112: groupBy() does not currently support grouping by related object fields
Hi! Big fan of SOQL Lib. We recently ran into a need to be able to use a `GROUP BY` clause with related object fields, which doesn't appear to be supported yet.

Instead of having to write:
```java
SOQL.of(OpportunityLineItem.SObjectType)
        .count(OpportunityLineItem.Name, 'count')
        .toString() + ' GROUP BY OpportunityLineItem.Opportunity.AccountId';
```

We'd like to be able to follow the "relationshipName" + SObjectField pattern:
```java
SOQL.of(OpportunityLineItem.SObjectType)
        .count(OpportunityLineItem.Name, 'count')
        .groupBy('OpportunityLineItem.Opportunity.Account', Account.Id);
```

I chose arbitrary SObjects for this example.

### PR DESCRIPTION
**Pull request by jessewheeler**
closes #112

### CODE REVIEW

File: force-app/main/default/classes/SOQL_Test.cls [hunk 1]
```diff
@@ -2519,4 +2519,26 @@ private class SOQL_Test {
         // Test
         SOQL.of(Account.SObjectType).count().preview().toInteger();
     }
+
+    @IsTest
```
Comment (pgajek2) [inline force-app/main/default/classes/SOQL_Test.cls:2523]: Hi @jessewheeler ! 

Your PR looks great!  🥳

Could we change the Unit Test a little bit? It's nice, but I would like to keep the same convention for all test methods. 

```java
    @IsTest
    static void groupByRelatedFormatting() {
        // Test
        String soql = SOQL.of(OpportunityLineItem.SObjectType)
            .count(OpportunityLineItem.Name, 'count')
            .groupBy('OpportunityLineItem.Opportunity.Account', Account.Id)
            .toString();

            // Verify
        Assert.areEqual(
            'SELECT COUNT(Name) count FROM OpportunityLineItem GROUP BY OpportunityLineItem.Opportunity.Account.Id',
            soql
        );
    }

    @IsTest
    static void groupByRelateToAggregated() {
        // Setup
        List<Account> accounts = insertAccounts();

        // Test
        List<AggregateResult> results = SOQL.of(Account.SObjectType)
            .count(Account.Name, 'names')
            .groupBy('Account.CreatedBy', User.Id)
            .toAggregated();

        // Verify
        Assert.areEqual(1, results.size());
    }
```

File: force-app/main/default/classes/SOQL.cls [hunk 2]
```diff
@@ -1589,6 +1600,11 @@ public virtual inherited sharing class SOQL implements Queryable {
             groupByFields.add(field.getDescribe().getName());
         }
 
+        public void with(String relationshipName, SObjectField field) {
+            setGroupByFunction('{0}');
+            groupByFields.add(relationshipName + '.' + field.getDescribe().getName());
```
Comment (pgajek2) [inline force-app/main/default/classes/SOQL.cls:1605]: We can do something like that, Apex invokes `.getDescribe().getName()` automatically.
It applies only when the String is concatenated with `SObjectField`.

`groupByFields.add(relationshipName + '.' + field);`

File: force-app/main/default/classes/SOQL.cls [hunk 3]
```diff
@@ -962,6 +969,10 @@ public virtual inherited sharing class SOQL implements Queryable {
             groupedFields.add(field.getDescribe().getName());
         }
 
+        public void withGroupedField(String relationshipName, SObjectField field) {
+            groupedFields.add(relationshipName + '.' + field.getDescribe().getName());
```
Comment (pgajek2) [inline force-app/main/default/classes/SOQL.cls:973]: We can do something like that, Apex invokes `.getDescribe().getName()` automatically.
It applies only when the String is concatenated with `SObjectField`.

`groupedFields.add(relationshipName + '.' + field);`

File: force-app/main/default/classes/SOQL_Test.cls (hunk 1 above)
Comment (jessewheeler) [reply force-app/main/default/classes/SOQL_Test.cls:2523]: Thanks, done!

**Pull request merged**